[v9] Support multiple transformations in role templates

vitorenesduarte opened the pull request:
Backport #19696 and #20412 to branch/v9

Commit:
[v9] Support multiple transformations in role templates

Before this commit, only a single transformation like `email.local` or
`regexp.replace` was allowed in role templates.

With this commit now users can write:
- `email.local(regexp.replace(external.email, "_", "-"))`
- `regexp.replace(email.local(regexp.replace(external.email, "_", "-")), "dev", "admin")`

This commit also allows the above expressions to not contain any
variable, i.e., it allows constant expressions like
`email.local("vitor@goteleport.com")`.

nklaassen commented on lib/utils/parse/ast.go:
```diff
@@ -0,0 +1,487 @@
+	// Kind indicates the expression kind.
+	Kind() reflect.Kind
+	// Evaluate evaluates the expression given the evaluation context.
+	Evaluate(ctx EvaluateContext) (any, error)
```
v9 is still on go1.17, there is no `any`, need to explicitly write `interface{}`

Commit:
any -> interface{}

marcoandredinis committed:
Return NotFound error when trait does not exist (#20412)

We were returning a trace.BadParameter, however we should return
trace.IsNotFound.
The godoc suggests that error.
The tests suggests that error, but they were only matching for type
which is the same for all the trace.XYZ errors.

Tests are now using the `trace.Is<Error>` helper method to correctly
validate the error message.

The side effect of this change is that we were receiving a lot of logs
when interpolating traits in a role.
That log is supressed when the error is NotFound.
That was not the case before this commit: the error was a BadParameter
one, so the log is written (as debug, but still a lot of noise).

Co-authored-by: Vitor Enes <vitor.duarte@goteleport.com>

Commit:
Merge branch 'branch/v9' into vitor/backport-19696-branch/v9

Commit:
Merge branch 'branch/v9' into vitor/backport-19696-branch/v9

Commit:
Merge branch 'branch/v9' into vitor/backport-19696-branch/v9
